File: Longest-Subarray-with-Sum-K.py
class Solution:
    def longestSubarray(self, arr, k):  
        # code here
        Lent=0
        sum=0
        nmap={0:1}
        
        for i in range(len(arr)):
            sum+=arr[i]
            if sum==k:
                Lent=max(Lent, i+1)
            if (sum-k) in nmap:
                Lent=max(Lent, i-nmap[sum-k])
            if not(sum in nmap):
                nmap[sum]=i
        return Lent
    

File: Longest Subarray with Sum K.js
class Solution {
    longestSubarray(arr, k) {
        let nmap = new Map(); // prefixSum -> first index
        let sum = 0;
        let Lent = 0;

        for (let i = 0; i < arr.length; i++) {
            sum += arr[i];

            if (sum === k) {
                Lent = Math.max(Lent, i + 1);
            }

            if (nmap.has(sum - k)) {
                Lent = Math.max(Lent, i - nmap.get(sum - k));
            }

            if (!nmap.has(sum)) {
                nmap.set(sum, i);
            }
        }

        return Lent;
    }
}
